### PR TITLE
Refactor Mills Equation: Align Enteric Methane Calculation with Mitscherlich Model 3 (Mills 2003)

### LINKED ISSUE #2341: [Animal] metabolizable_energy_intake Incorrectly Traced, Causing Zero Values in Enteric Methane Calculation
## Expected Behavior  
The `metabolizable_energy_intake` value should be correctly retrieved and used when calculating **enteric methane emissions** using the **Mills equation** for both **lactating and dry cows**.

## Current Behavior  
The code currently **does not trace or retrieve the `metabolizable_energy_intake` variable properly**, resulting in a value of **0**.  

Image

As a result, **enteric methane emissions are incorrectly calculated as 0** when using the **Mills equation**, despite valid diet and animal input.

## Possible Solution  
- Fix the code to ensure that `metabolizable_energy_intake` is correctly traced and passed to the enteric methane calculation for all relevant animal categories.
- Verify that the traced value is **non-zero and biologically reasonable** before performing the methane emission calculation.

## Steps to Reproduce  
1. Run a simulation with default inputs using the **Mills equation** to calculate enteric methane.  
2. Observe that the enteric methane output for lactating and dry cows is 0, despite the presence of valid dietary inputs.
3. Print `metabolizable_energy_intake` values and they are 0.

## Context (Environment)  
This issue affects **methane emission estimates** for lactating and dry cows and leads to **underreporting of emissions** when using the Mills method.

## Detailed Description  
The current implementation fails to correctly trace the value of `metabolizable_energy_intake`, leading to **incorrect methane estimates (zero)** when applying the Mills equation. This disrupts the integrity of model results for lactating and dry cow groups, particularly when ME intake is a required input for methane calculations.

## Possible Implementation  
- Trace the correct variable from the relevant dietary or intake outputs.
- Ensure it is mapped properly for each animal group before passing it into the Mills equation logic.
- Add a test to validate non-zero `metabolizable_energy_intake` values during simulation runs.

## Target Deadline  
Recommend addressing this in the next sprint to restore accurate methane emission calculations for lactating and dry cows.

### PR DESCRIPTION
**Pull request by tomhuhh**
This pull request updates the code for the Mills equation to calculate enteric methane emissions from lactating and dry cows, aligning it with the structure of the Mitscherlich Model 3 as described in Mills (2003) 
 <https://doi.org/10.2527/2003.81123141x>.

This pull request now addresses issue #2341. (Updated on April 16th) 

### Context

Screenshot 2025-04-14 at 3 48 07 PM

### What

- Updated the existing Mills equation implementation to better represent the structure of the Mitscherlich Model 3.
- Incorporated constants in the animal_constants.py and general_constants.py files to streamline the code. 
- Scientific documentation has been revisited and no changes are needed.
- There is no need for changes in unit tests as the update does not change the original function of the Mills equation.

### Why

- The previous implementation did not reflect the intended structure of the Mitscherlich Model 3, potentially leading to misunderstanding of the methane emission code.
- These modifications are crucial for ensuring the quality and readability of the enteric methane emission code.

### How

- Conducted a thorough review of Mills (2003) to extract the key aspects of the Mitscherlich Model 3.
- Refactored the equation's logic to closely align with the structure of Mitscherlich Model 3.
- Performed code reviews and manual checks to ensure that the new implementation performs as intended.

### Test plan

Use the filter below: 
`AnimalModuleReporter.report_enteric_methane_emission.enteric_methane_emission_for_pen_.*`

### CODE REVIEW

File: RUFAS/biophysical/animal/animal_constants.py [hunk 1]
```diff
@@ -111,3 +111,9 @@
 # The buying and selling threshold of for herd-size maintenance
 BUYING_THRESHOLD = 1.01
 SELLING_THRESHOLD = 1.03
+
+# The parameter a in Mitscherlich Model 3 (Mills et al., 2003)
+MITS_PARAMETER_A = 45.98
+
+# The parameter b in Mitscherlich Model 3 (Mills et al., 2003)
+MITS_PARAMETER_B = 0
```
Comment (matthew7838) [inline RUFAS/biophysical/animal/animal_constants.py]: I don't think this is a big issue, but we should try to avoid inline comments. By looking at these values, I think it's worth it to make them a one-line docstring description.

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py [hunk 2]
```diff
@@ -75,7 +78,7 @@ def calculate_heifer_methane(
             )
             methane_emission = (
                 0.065 * gross_energy_concentration * nutrition_supply.dry_matter
-            ) / 0.05565  # [A.3B.C.3]
+            ) / GeneralConstants.MJ_CH4_TO_G_CH4  # [A.3B.C.3]
```
Comment (matthew7838) [inline RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: We should move the reference in the docstring reference section.

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py [hunk 3]
```diff
@@ -26,7 +27,9 @@ def calculate_calf_methane(methane_model: str | None, body_weight: float) -> flo
         """
         methane_emission = 0.0
         if methane_model:
-            methane_emission = (0.013 * (body_weight**0.75) * 4.184) / 0.05565
+            methane_emission = (
+                0.013 * (body_weight**0.75) * GeneralConstants.CALORIES_TO_JOULES
```
Comment (jadamchick) [inline RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: is there a purpose to the double ** here or is that a typo?

File: RUFAS/general_constants.py [hunk 4]
```diff
@@ -75,6 +75,8 @@ class GeneralConstants:
     MILK_FAT_WEIGHT = 12.2
     FRACTION_OF_HUMIC_NITROGEN_IN_ACTIVE_POOL = 0.02
     """Defined in SWAT Theoretical documentation, page 186 in paragraph beneath eqn. 3:1.1.4."""
+    CALORIES_TO_JOULES = 4.184
```
Comment (jadamchick) [inline RUFAS/general_constants.py]: it looks as though you are usually converting kilocalories to megajoules with this conversion (and then following it by converting MJ methane to g methane) -- it might be slightly clearer if this was defined as kcal_to_MJ?  just an idea

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py (hunk 2 above)
Comment (jadamchick) [inline RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: equation reference in comments rather than notes/references -- also, should this be updated to the new naming format?

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py [hunk 5]
```diff
@@ -293,22 +298,22 @@ def _calculate_dry_cow_enteric_methane(
...
             gross_energy_concentration = (
                 0.263 * crude_protein_concentration
                 + 0.522 * ethyl_ester_concentration
                 + 0.198 * neutral_detergent_fiber_concentration
                 + 0.160 * soluble_residue
             )
-            methane_emission = (0.065 * gross_energy_concentration * dry_matter_intake) / 0.05565  # [A.3B.C.3]
+            methane_emission = (
+                0.065 * gross_energy_concentration * dry_matter_intake
+            ) / GeneralConstants.MJ_CH4_TO_G_CH4  # [A.3B.C.3]
```
Comment (jadamchick) [inline RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: equation reference -- see comment above

**Review by jadamchick — APPROVED** [review]
Hi Haowen,
overall this looks great, very clearly explained and documented. I left a few minor comments on ways to possibly add clarity to the code. 

I ran the default herd with IPCC and then Mills methane model specified. The only values that are different between outputs (that I could detect) are the enteric methane for close-up and lact cow pens. Full results are here:
[enteric methane mills v. ipcc.xlsx](https://github.com/user-attachments/files/19914288/enteric.methane.mills.v.ipcc.xlsx)
The close-up pen has higher (more) methane calculated by Mills than by IPCC for all days.
The lact cow pen has higher methane calculated by Mills for the first 31 days (presumably until the ration is reformulated) and then has a higher value calculated by IPCC for the rest of the simulation. 

does this reflect what you expect to occur? assuming yes, it looks good to me.

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py (hunk 3 above)
Comment (tomhuhh) [reply RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: Hi Julie, the `**` operator here represents exponentiation.

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py (hunk 2 above)
Comment (tomhuhh) [reply RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: Great point! I’ve removed the inline comments, but I believe Kristan’s branch updating the new equation ID convention hasn’t been merged into dev yet—so this file is still using the older convention.

**Comment by tomhuhh** [comment]
Thanks for your review, @jadamchick! Since the Mits3 model was developed for lactating and dry cows only, it makes sense that you’re seeing differences in the close-up and lactating pens. And for the same reason, enteric methane will be higher in the lactating pen compared to the close-up pen, which also includes heifers.

**Comment by jadamchick** [comment]
> Thanks for your review, @jadamchick! Since the Mits3 model was developed for lactating and dry cows only, it makes sense that you’re seeing differences in the close-up and lactating pens. And for the same reason, enteric methane will be higher in the lactating pen compared to the close-up pen, which also includes heifers.

Makes sense. To be clear, I was also comparing lactating cows with IPCC to lactating cows with Mills. For the first 31 days, the Mills prediction was higher than IPCC. For the remaining thousands of days, the IPCC prediction was higher than Mills. I presume this is associated with a ration reformulation after the first 30 days, and the specifics of the ration being fed determine which method predicts a higher methane? But I'm not enough of an expert to verify that assumption :)

**Comment by tomhuhh** [comment]
Thanks @jadamchick! the change between the first 30 days and later is driven by the increase in dry matter intake. The consistent relative difference in enteric CH₄ predictions between the two models likely reflects their different methodologies.

**Review by matthew7838 — APPROVED** [review]
LGTM, have really minor suggestions. Sorry for the delay.

File: RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py (hunk 2 above)
Comment (tomhuhh) [reply RUFAS/biophysical/animal/digestive_system/enteric_methane_calculator.py]: Thanks Matthew! I think this has already been fixed.
